3DS Backend

Cruel opened the pull request:
I've been testing this native Nintendo 3DS backend for about a week both myself and with a couple [unofficial test releases the 3ds homebrew community](https://gbatemp.net/threads/release-scummvm-3ds.422051/).

If you want to build it and don't know about 3ds toolchain or dependencies, I got the basic information needed [here](https://github.com/Cruel/scummvm/wiki/Compiling). Please review and let me know if I screwed up anything lol.

criezy commented:
I know nothing about the 3DS and I can't really comment on all the changes (other than say this is always great to add support for a new platform - so thank you).

But I will comment on something that has nothing todo with the backend code: the proper way to add yourself to the credits is to modify devtools/credits.pl and then run 'make credits'. Unless I missed something it looks like you modified directly gui/credits.h (and didn't modify the other credits files).

Cruel commented:
Ah, didn't see that. At least I modified the header properly lol, the script generated the same output.

sev- commented on backends/platform/3ds/osystem-audio.cpp:
```diff
@@ -0,0 +1,106 @@
+	uint32 time = lastTime;
+	ndspWaveBuf buffers[bufferCount];
+	
+	for(i = 0; i < bufferCount; ++i) {
```
Space after 'for' is missing.

sev- commented on backends/platform/3ds/osystem-audio.cpp:
```diff
@@ -0,0 +1,106 @@
+	ndspChnSetRate(channel, sampleRate);
+	ndspChnSetFormat(channel, NDSP_FORMAT_STEREO_PCM16);
+
+	while(!osys->exiting) {		
```
Same thing, reserved words must have space after them:

 http://wiki.scummvm.org/index.php/Code_Formatting_Conventions#Whitespaces

sev- commented on backends/platform/3ds/osystem-audio.cpp:
```diff
@@ -0,0 +1,106 @@
+		if (!osys->sleeping && sampleLen > 0) {
+			bufferIndex++;
+			bufferIndex %= bufferCount;
+			ndspWaveBuf* buf = &buffers[bufferIndex];
```
ndspWaveBuf *buf

sev- commented on backends/platform/3ds/osystem-audio.cpp:
```diff
@@ -0,0 +1,106 @@
+}
+
+void OSystem_3DS::initAudio() {
+	_mixer = new Audio::MixerImpl(this, 22050);
```
Does the console able to work with other sample rates? We have a command-line parameter '--output-rate=RATE', which is also presented in config file. So, if it is able to have more rates like 44100, it makes sense to support that rather hardcoding.

Cruel replied:
Good point. Accepting command-line args on the 3DS is partially supported, but it is a complication I left out for now. Once figured out, I will refactor the things influenced by the args. It requires manually creating an XML file shortcut on the SD card specifying args, and it's not widely used.

sev- commented on backends/platform/3ds/osystem-audio.cpp:
```diff
@@ -0,0 +1,106 @@
+	if (hasAudio) {
+		s32 prio = 0;
+		svcGetThreadPriority(&prio, CUR_THREAD_HANDLE);
+		audioThread = threadCreate(&audioThreadFunc, _mixer, 32*1048, prio-1, -2, false);
```
32 * 1048, prio - 1,

sev- commented on backends/platform/3ds/osystem-events.cpp:
```diff
@@ -0,0 +1,231 @@
+}
+
+static void eventThreadFunc(void* arg) {
+	OSystem_3DS* osys = (OSystem_3DS*) g_system;
```
OSystem_3DS *osys = (OSystem_3DS *)g_system;

etc in similar cases

sev- commented on backends/platform/3ds/osystem-events.cpp:
```diff
@@ -0,0 +1,231 @@
+	auto eventQueue = (Common::Queue<Common::Event>*) arg;
+	
+	uint32 touchStartTime = osys->getMillis();
+	touchPosition lastTouch = {0,0};
```
{0, 0};

sev- commented on backends/platform/3ds/osystem-events.cpp:
```diff
@@ -0,0 +1,231 @@
+			}
+			
+			lastTouch = touch;
+		}
```
} else if {

sev- commented on backends/platform/3ds/osystem-events.cpp:
```diff
@@ -0,0 +1,231 @@
+		else if (keysReleased & KEY_TOUCH) {
+			event.mouse.x = lastTouch.px;
+			event.mouse.y = lastTouch.py;
+			printf("clicked %u, %u\n", lastTouch.px, lastTouch.py);
```
Is this a leftover?

sev- commented on backends/platform/3ds/osystem-events.cpp:
```diff
@@ -0,0 +1,231 @@
+				pushEventQueue(eventQueue, event);
+			}
+		}
+		if (keysPressed & KEY_R) {
```
It might make sense to support GUI_KEYS_DIALOG, like Symbian and WinCE do, or alternatively, employ KeyMapper, instead of hardcoding the key meanings.

sev- commented on backends/platform/3ds/osystem-graphics.cpp:
```diff
@@ -0,0 +1,474 @@
+bool OSystem_3DS::hasFeature(OSystem::Feature f) {
+	return (f == OSystem::kFeatureFullscreenMode ||
+	        f == OSystem::kFeatureCursorPalette ||
+	        f == OSystem::kFeatureOverlaySupportsAlpha);
```
Maybe it will be obvious further, but why don't you use the VirtualKeyboard?

sev- commented on backends/platform/3ds/osystem.cpp:
```diff
@@ -0,0 +1,176 @@
+#include <time.h>
+
+OSystem_3DS::OSystem_3DS():
+_focusDirty(true),
```
No indentation?

sev- commented on backends/platform/3ds/shader.v.pica:
```diff
@@ -0,0 +1,40 @@
+; PICA200 vertex shader
```
What is the copyright of this file?

sev- commented on backends/platform/3ds/sprite.cpp:
```diff
@@ -0,0 +1,144 @@
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+#define FORBIDDEN_SYMBOL_ALLOW_ALL
```
Why do you need it in this file?

sev- commented:
Overall it looks decent, and as I understand, you're willing to join the team and continue supporting your port, right?

Besides the usual problems with our code formatting requirements, I have couple of questions:

1. The port lacks the port-specific README. E.g. instructions on the keys, where to put games, how to put ScummVM on the device etc.
2. You're not using the virtual keyboard. How the users are supposed to enter save names?
3. What is the purpose of banner.wav file?
4. You might consider using either GUI_ENABLE_KEYSDIALOG #define, or --enable-keymapper, so the users are able to remap the keys if needed.

Cruel commented:
Yes, I'm willing to maintain/support this port. To address your questions:

1. Added the README just now.
2. I have a key defined that opens the integrated virtual keyboard for now. I was going to use the value provided by `setFeatureState` to show/hide the keyboard, but with a lack of direct control over the keyboard (beyond issuing an Event to toggle it, blindly unaware of whether it was showing or not), was problematic. My best ideas to solve this required some restructuring of the keyboard event or the default event manager in some minor way, so I held off on that for now to perhaps discuss that elsewhere.
3. The banner.wav is currently an empty audio file needed to create a 3DS banner. I also have a question regarding this. [Here a video demonstrating what 3DS banners are](https://www.youtube.com/watch?v=P6x2SKLx7Rc). As seen in the video, there are two different banners made for ScummVM. I have only included in the repo the plain one with empty audio and ScummVM logo. I would like to include the 3D banner with audio and animation however I'm unaware of copyright details with any depicted characters (though I can obviously make sure the audio is free of copyright).
4. Using the key mapper is on my TODO list, right after I finish adding a GUI dialog for 3DS specific configurations like disabling the top/bottom screen.
5. I'll keep this in mind. I removed them now. My usage of it was mostly left over from when I was using printf and other functions during testing.

Everything was built using linux, so I can help with the buildbot when needed.

sev- commented:
Great! Please come to IRC for discussion on how to join the team.

Also I wonder, can you make the port for 1.8.1? That means committing same changes to branch-1-8 by this Friday, 20th of May and prepare the binaries after this date.

BeeMovieScript commented:
The 3DS version doesn't support games designed for 640x480 such as most Humongous Entertainment games
